src/main/java/fi/aalto/cs/apluscourses/intellij/utils/Interfaces.java

```diff
@@ -88,6 +114,133 @@ public static void refreshPath(VirtualFile file, Runnable callback) {
+        jsonObject.put(JSON_ENTRY_NAME, new JSONArray(hashes));
+
+        return jsonObject;
+      }
+
+      public boolean containsHash(@NotNull String hash) {
+        return hashes.contains(hash);
+      }
+
+      public void addHash(@NotNull String hash) {
+        hashes.add(hash);
+      }
+    }
+
+    private static Cache<String, JSONObject> submissionsCache;
+
+    private DuplicateSubmissionCheckerImpl() {
+
+    }
+
+    private static String hashFileToString(@NotNull MessageDigest digest, @NotNull Path path) {
+      try {
+        byte[] hash = digest.digest(Files.readAllBytes(path));
+        return Base64.getEncoder().encodeToString(hash);
+      } catch (IOException e) {
+        throw new UncheckedIOException(e);
+      }
+    }
+
+    /**
+     * Checks whether the same submission has already been submitted by the user. If it is the case, the user
+     * is asked whether they want to submit anyway or abort the submission.
+     * @param project The current IntelliJ project.
+     * @param courseId ID of the course which contains the exercise in question.
+     * @param exerciseId ID of the exercise.
+     * @param files A map of submittable files, as constructed in {@link SubmitExerciseAction}.
+     * @return True if duplicate check succeeded and the submission should proceed; false if it should be cancelled.
+     */
+    public static boolean checkForDuplicateSubmission(@NotNull Project project,
+                                                      @NotNull String courseId,
+                                                      long exerciseId,
+                                                      @NotNull Map<String, Path> files) {
+      // we want to defer creation of the cache until it's actually necessary
+      if (submissionsCache == null) {
+        Path projectPath = Path.of(Objects.requireNonNull(project.getBasePath()));
+        submissionsCache = new JsonFileCache(
+            projectPath.resolve(Path.of(Project.DIRECTORY_STORE_FOLDER, "a-plus-hashes.json")));
+      }
+
+      try {
+        final MessageDigest shaDigest = MessageDigest.getInstance("SHA-256");
+        final StringBuilder submissionString = new StringBuilder();
+
+        // the iteration order is important so that files always get concatenated in the same order
+        files.entrySet().stream().sorted(Map.Entry.comparingByKey()).forEachOrdered(entry -> {
+          submissionString.append(entry.getKey());
+          submissionString.append('|'); // a separating character that is not in base-64 alphabet
+          submissionString.append(hashFileToString(shaDigest, entry.getValue()));
+          submissionString.append(',');
+        });
+
+        final String cacheKey = "hash_c" + courseId + "_e" + exerciseId;
+        final byte[] finalHashBytes = shaDigest.digest(submissionString.toString().getBytes(StandardCharsets.UTF_8));
+        final String finalHash = Base64.getEncoder().encodeToString(finalHashBytes);
```

Comment:
I guess this is ok. I would perhaps find it more intuitive if each file was first hashed (`<key>:<content>` to SHA256) and these hashes were XORred to make the final hash; that way the order wouldn't matter either. Hashing hashes always raises a concern of a diminishing domain space but in this case, it is really a non-issue because the number of hashes are so small.